core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -0,0 +1,319 @@
+                           epoch: Short,
+                           command: TransactionResult,
+                           responseCallback: EndTxnCallback): Unit = {
+
+    val errors = validateTransactionalId(transactionalId)
+    if (errors != Errors.NONE)
+      responseCallback(errors)
+    else
+      txnManager.getTransaction(transactionalId) match {
+        case None =>
+          responseCallback(Errors.INVALID_PID_MAPPING)
+        case Some(metadata) =>
+          metadata synchronized {
+            if (metadata.pid != pid)
+              responseCallback(Errors.INVALID_PID_MAPPING)
+            else if (metadata.epoch != epoch)
+              responseCallback(Errors.INVALID_PRODUCER_EPOCH)
+            else metadata.state match {
+              case Ongoing =>
+                commitOrAbort(transactionalId, pid, epoch, command, responseCallback, metadata)
+              case CompleteCommit =>
+                if (command == TransactionResult.COMMIT)
+                  responseCallback(Errors.NONE)
+                else
+                  responseCallback(Errors.INVALID_TXN_STATE)
+              case CompleteAbort =>
+                if (command == TransactionResult.ABORT)
+                  responseCallback(Errors.NONE)
+                else
+                  responseCallback(Errors.INVALID_TXN_STATE)
+              case _ =>
+                responseCallback(Errors.INVALID_TXN_STATE)
+            }
+          }
+      }
+  }
+
+  private def commitOrAbort(transactionalId: String,
+                            pid: Long,
+                            epoch: Short,
+                            command: TransactionResult,
+                            responseCallback: EndTxnCallback,
+                            metadata: TransactionMetadata) = {
+    val nextState = if (command == TransactionResult.COMMIT) PrepareCommit else PrepareAbort
+    val newMetadata = new TransactionMetadata(pid, epoch, metadata.txnTimeoutMs, nextState, metadata.topicPartitions, time.milliseconds())
+    metadata.prepareTransitionTo(nextState)
+
+    def logAppendCallback(errors: Errors): Unit = {
+      if (errors != Errors.NONE)
+        responseCallback(errors)
+      else {
+        txnManager.coordinatorEpochFor(transactionalId) match {
+          case Some(coordinatorEpoch) =>
+            def completionCallback(): Unit = {
+              val preparedCommitMetadata = txnManager.getTransaction(transactionalId).get
+              val completedState = if (nextState == PrepareCommit) CompleteCommit else CompleteAbort
+              val committedMetadata = new TransactionMetadata(pid,
+                epoch,
+                newMetadata.txnTimeoutMs,
+                completedState,
+                metadata.topicPartitions,
+                newMetadata.timestamp)
+              preparedCommitMetadata.prepareTransitionTo(completedState)
+              txnManager.appendTransactionToLog(transactionalId, committedMetadata, responseCallback, 1)
```

Comment:
According to point 2 https://docs.google.com/document/d/11Jqy_GjUGtdXJK94XGsEIK7CP1SnQGdp2eF0wSw9ra8/edit#bookmark=id.3af5934pfogc

>When all the responses have been received, append a COMPLETE_COMMIT transaction message to the transaction topic. We do not need to wait for this record to be fully replicated since otherwise we will just redo this protocol again.